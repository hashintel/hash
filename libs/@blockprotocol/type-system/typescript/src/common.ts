// import init from "@blockprotocol/type-system-rs";

// export type InitInput =
//   | RequestInfo
//   | URL
//   | Response
//   | BufferSource
//   | WebAssembly.Module;
//
// let wasmInit: (() => InitInput) | undefined = undefined;
// export const setWasmInit = (arg: () => InitInput) => {
//   wasmInit = arg;
// };
//
// let initialized: Promise<void> | undefined = undefined;
// export class TypeSystemInitializer {
//   private constructor() {}
//
//   /**
//    * Initializes the package. There is a one time global setup fee (sub 30ms), but subsequent
//    * requests to initialize will be instantaneous, so it's not imperative to reuse the same parser.
//    */
//   public static initialize = async (options?: InitInput) => {
//     if (initialized === undefined) {
//       const loadModule = options ?? (wasmInit ? wasmInit() : undefined);
//       initialized = init(loadModule).then(() => undefined);
//     }
//
//     await initialized;
//     return new TypeSystemInitializer();
//   };
// }

/**
 * Ensures that the array has at least one element. If not it returns `undefined`.
 *
 * @todo: Remove when Typescript can infer this
 * @see https://github.com/microsoft/TypeScript/issues/29841
 */
export const atLeastOne = <T>(array: T[]): [T, ...T[]] | undefined =>
  // @ts-expect-error –– @see https://github.com/microsoft/TypeScript/issues/29841
  array.length > 0 ? (array satisfies [T, ...T[]]) : undefined;
