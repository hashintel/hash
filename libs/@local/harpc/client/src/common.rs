//! Common utilities used to implement various traits as well as used in the proc-macro.

use core::error::Error;

use error_stack::{Report, TryReportStreamExt as _};
use futures::stream;
use harpc_codec::encode::Encoder;
use harpc_net::session::server::SessionId;
use harpc_service::Service;
use harpc_tower::{
    Extensions,
    request::{self, Request},
};
use harpc_types::{procedure::ProcedureDescriptor, service::ServiceDescriptor};

use crate::connection::ConnectionRequestStream;

/// Encode a request of an iterator of items.
pub async fn encode_request_iter<P, E, C>(
    codec: E,
    procedure: P,
    items: impl IntoIterator<Item: serde::Serialize, IntoIter: Send> + Send,
) -> Result<Request<ConnectionRequestStream<E>>, Report<[C]>>
where
    P: harpc_service::procedure::ProcedureIdentifier + Send,
    E: Encoder<Error = Report<C>, Buf: Send> + Send,
    C: Error + Send + Sync + 'static,
{
    let items: Vec<_> = codec
        .encode(stream::iter(items))
        .try_collect_reports()
        .await?;

    Ok(Request::from_parts(
        request::Parts {
            service: ServiceDescriptor {
                id: <P::Service as Service>::ID,
                version: <P::Service as Service>::VERSION,
            },
            procedure: ProcedureDescriptor {
                id: procedure.into_id(),
            },
            session: SessionId::CLIENT,
            extensions: Extensions::new(),
        },
        stream::iter(items),
    ))
}
