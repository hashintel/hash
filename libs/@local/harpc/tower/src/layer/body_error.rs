use core::{
    error::Error,
    task::{Context, Poll},
};

use futures::TryFutureExt;
use harpc_codec::encode::ErrorEncoder;
use harpc_types::response_kind::ResponseKind;
use tower::{Layer, Service};

use crate::{
    body::{Body, encode_error::EncodeError},
    request::Request,
    response::Response,
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct HandleBodyErrorLayer<E> {
    encoder: E,
}

impl<E> HandleBodyErrorLayer<E> {
    pub const fn new(encoder: E) -> Self {
        Self { encoder }
    }
}

impl<E, S> Layer<S> for HandleBodyErrorLayer<E>
where
    E: Clone,
{
    type Service = HandleBodyErrorService<S, E>;

    fn layer(&self, inner: S) -> Self::Service {
        HandleBodyErrorService {
            inner,
            encoder: self.encoder.clone(),
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct HandleBodyErrorService<S, E> {
    inner: S,

    encoder: E,
}

impl<S, E, ReqBody, ResBody> Service<Request<ReqBody>> for HandleBodyErrorService<S, E>
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send,
    E: ErrorEncoder + Clone,
    ReqBody: Body<Control = !>,
    // the extra bounds here are not strictly required, but they help to make the error messages
    // more expressive during compilation
    ResBody: Body<Control: AsRef<ResponseKind>, Error: Error + serde::Serialize>,
{
    type Error = S::Error;
    type Response = Response<EncodeError<ResBody, E>>;

    type Future = impl Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        let encoder = self.encoder.clone();

        self.inner
            .call(req)
            .map_ok(|res| res.map_body(|body| EncodeError::new(body, encoder)))
    }
}
