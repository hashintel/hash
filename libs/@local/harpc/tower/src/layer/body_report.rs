use core::task::{Context, Poll};

use error_stack::Report;
use futures::TryFutureExt as _;
use harpc_types::response_kind::ResponseKind;
use tower::{Layer, Service};

use crate::{
    body::{Body, encode_report::EncodeReport},
    request::Request,
    response::Response,
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct HandleBodyReportLayer {
    _private: (),
}

impl HandleBodyReportLayer {
    #[expect(
        clippy::new_without_default,
        reason = "layer construction should be explicit and we might add fields in the future"
    )]
    #[must_use]
    pub const fn new() -> Self {
        Self { _private: () }
    }
}

impl<S> Layer<S> for HandleBodyReportLayer {
    type Service = HandleBodyErrorService<S>;

    fn layer(&self, inner: S) -> Self::Service {
        HandleBodyErrorService { inner }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct HandleBodyErrorService<S> {
    inner: S,
}

impl<S, C, ReqBody, ResBody> Service<Request<ReqBody>> for HandleBodyErrorService<S>
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send,
    ReqBody: Body<Control = !>,
    // The extra bounds here are not strictly required, but they help to make the error messages
    // more expressive during compilation
    ResBody: Body<Control: AsRef<ResponseKind>, Error = Report<C>>,
    C: error_stack::Context,
{
    type Error = S::Error;
    type Response = Response<EncodeReport<ResBody>>;

    type Future = impl Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        self.inner
            .call(req)
            .map_ok(|res| res.map_body(|body| EncodeReport::new(body)))
    }
}
