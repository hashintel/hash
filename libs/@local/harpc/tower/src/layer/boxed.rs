use core::task::{Context, Poll};

use bytes::Buf;
use harpc_wire_protocol::response::kind::ResponseKind;
use tower::{Layer, Service, ServiceExt};

use crate::{
    body::{Body, BodyExt},
    request::Request,
    response::{BoxedResponse, Response},
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct BoxedResponseLayer {
    _private: (),
}

impl BoxedResponseLayer {
    #[must_use]
    pub const fn new() -> Self {
        Self { _private: () }
    }
}

impl<S> Layer<S> for BoxedResponseLayer {
    type Service = BoxedResponseService<S>;

    fn layer(&self, inner: S) -> Self::Service {
        BoxedResponseService { inner }
    }
}

impl Default for BoxedResponseLayer {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct BoxedResponseService<S> {
    inner: S,
}

impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for BoxedResponseService<S>
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>, Error = !> + Clone + Send,
    ReqBody: Body<Control = !>,
    ResBody: Body<Control: AsRef<ResponseKind>> + Send + Sync + 'static,
{
    type Error = !;
    // TODO: do we wanna put the response error into a dyn error?!
    type Response = BoxedResponse<ResBody::Error>;

    type Future = impl Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        // Taken from axum::HandleError
        // we're always ready because we clone the inner service, therefore it is unused and always
        // ready
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        let clone = self.inner.clone();
        let inner = core::mem::replace(&mut self.inner, clone);

        async move {
            let Ok(response) = inner.oneshot(req).await;

            let response = response.map_body(|body| {
                body.map_control(|control| *control.as_ref())
                    .map_data(|mut buffer| {
                        let length = buffer.remaining();
                        buffer.copy_to_bytes(length)
                    })
                    .boxed()
            });

            Ok(response)
        }
    }
}
