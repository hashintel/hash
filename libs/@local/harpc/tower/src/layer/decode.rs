use core::{
    marker::PhantomData,
    mem,
    task::{Context, Poll},
};

use bytes::Buf;
use futures::Stream;
use harpc_codec::decode::Decoder;
use tower::{Layer, Service};

use crate::{request::Request, response::Response};

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct DecodeLayer<C, T> {
    codec: C,
    _marker: PhantomData<fn() -> *const T>,
}

impl<C> DecodeLayer<C, !> {
    pub const fn new<T>(codec: C) -> DecodeLayer<C, T> {
        DecodeLayer {
            codec,
            _marker: PhantomData,
        }
    }
}

impl<S, C, T> Layer<S> for DecodeLayer<C, T>
where
    C: Clone,
{
    type Service = DecodeService<S, C, T>;

    fn layer(&self, inner: S) -> Self::Service {
        DecodeService {
            inner,
            codec: self.codec.clone(),
            _marker: PhantomData,
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct DecodeService<S, C, T> {
    inner: S,
    codec: C,
    _marker: PhantomData<fn() -> *const T>,
}

impl<S, C, T, ReqBody, ResBody, ResData, ResError> Service<Request<ReqBody>>
    for DecodeService<S, C, T>
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone,
    ResBody: Stream<Item = Result<ResData, ResError>> + Send,
    C: Decoder + Clone,
    ResData: Buf,
    T: serde::de::DeserializeOwned,
{
    type Error = S::Error;
    type Response = Response<<C as Decoder>::Output<T, ResBody>>;

    type Future = impl Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        let codec = self.codec.clone();

        // see: https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
        let clone = self.inner.clone();
        let mut inner = mem::replace(&mut self.inner, clone);

        async move {
            let response = inner.call(req).await?;
            let (parts, body) = response.into_parts();

            let stream = codec.decode(body);

            Ok(Response::from_parts(parts, stream))
        }
    }
}

// TODO: test
