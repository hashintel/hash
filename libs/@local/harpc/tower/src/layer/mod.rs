pub mod error;
pub mod report;
