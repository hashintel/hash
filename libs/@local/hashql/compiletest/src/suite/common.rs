use core::fmt::{self, Display, Write as _};

use hashql_core::span::SpanId;
use hashql_diagnostics::{Diagnostic, category::DiagnosticCategory};

use super::SuiteDiagnostic;

/// Process diagnostics from a lowering step.
///
/// This will scan through all the reported diagnostics and
/// push them into the output vector. If a fatal diagnostic is found,
/// it will be returned as an error.
pub(crate) fn process_diagnostics<C>(
    output: &mut Vec<SuiteDiagnostic>,
    reported: impl IntoIterator<Item = Diagnostic<C, SpanId>>,
) -> Result<(), SuiteDiagnostic>
where
    C: DiagnosticCategory + 'static,
{
    let mut fatal = None;

    for diagnostic in reported {
        let diagnostic = diagnostic.boxed();

        if fatal.is_none() && diagnostic.severity.is_fatal() {
            fatal = Some(diagnostic);
            continue;
        }

        output.push(diagnostic);
    }

    fatal.map_or(Ok(()), Err)
}

#[track_caller]
pub(crate) fn process_result<C, T, I>(
    output: &mut Vec<SuiteDiagnostic>,
    result: Result<T, I>,
) -> Result<T, SuiteDiagnostic>
where
    I: IntoIterator<Item = Diagnostic<C, SpanId>>,
    C: DiagnosticCategory + 'static,
{
    match result {
        Ok(value) => Ok(value),
        Err(diagnostics) => {
            let diagnostic = process_diagnostics(output, diagnostics)
                .expect_err("reported diagnostics should always be fatal");

            Err(diagnostic)
        }
    }
}

pub(crate) struct Header(&'static str);

impl Header {
    pub(crate) const fn new(title: &'static str) -> Self {
        Self(title)
    }
}

impl Display for Header {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        const HEADER: char = '\u{2550}';

        let len = self.0.len();

        for _ in 0..4 {
            fmt.write_char(HEADER)?;
        }

        write!(fmt, " {} ", self.0)?;

        let remaining = (80_usize - 4).saturating_sub(len + 2);

        for _ in 0..remaining {
            fmt.write_char(HEADER)?;
        }

        Ok(())
    }
}

pub(crate) struct Annotated<T, U> {
    pub content: T,
    pub annotation: U,
}

impl<T, U> Display for Annotated<T, U>
where
    T: Display,
    U: Display,
{
    #[expect(clippy::non_ascii_literal)]
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
        struct Prefix {
            start: &'static str,
            continuation: &'static str,
        }

        impl Prefix {
            const ANNOTATION: Self = Self {
                start: "└→ ",
                continuation: "   ",
            };
            const CONTENT: Self = Self {
                start: "┌─ ",
                continuation: "│  ",
            };

            const fn select(self, line: usize) -> &'static str {
                if line == 0 {
                    self.start
                } else {
                    self.continuation
                }
            }
        }

        struct Writer<'a, 'b> {
            inner: &'a mut fmt::Formatter<'b>,
            prefix: Prefix,

            line_number: usize,
            pending: bool,
        }

        impl<'a, 'b> Writer<'a, 'b> {
            const fn new(inner: &'a mut fmt::Formatter<'b>, prefix: Prefix) -> Self {
                Self {
                    inner,
                    prefix,

                    line_number: 0,
                    pending: true,
                }
            }

            const fn reset(&mut self, prefix: Prefix) {
                self.prefix = prefix;
                self.line_number = 0;
                self.pending = true;
            }
        }

        impl fmt::Write for Writer<'_, '_> {
            #[expect(clippy::renamed_function_params)]
            fn write_str(&mut self, string: &str) -> fmt::Result {
                let lines = string.split_inclusive('\n');
                for line in lines {
                    if self.pending {
                        self.inner.write_str(self.prefix.select(self.line_number))?;
                    }

                    self.inner.write_str(line)?;

                    if line.ends_with('\n') {
                        self.line_number += 1;
                        self.pending = true;
                    } else {
                        self.pending = false;
                    }
                }

                Ok(())
            }
        }

        let options = fmt.options();

        let mut writer = Writer::new(fmt, Prefix::CONTENT);

        let mut formatter = fmt::Formatter::new(&mut writer, options);
        self.content.fmt(&mut formatter)?;
        writer.write_str("\n")?;

        writer.reset(Prefix::ANNOTATION);

        let mut formatter = fmt::Formatter::new(&mut writer, options);
        self.annotation.fmt(&mut formatter)?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    #![expect(clippy::non_ascii_literal)]
    use core::fmt;

    use crate::suite::common::Annotated;

    struct MultipleWrites;

    impl fmt::Display for MultipleWrites {
        fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
            fmt.write_str("foo")?;
            fmt.write_str("bar")?;
            fmt.write_str("hello\nworld")
        }
    }

    /// A simple single-line type for the annotation
    struct SimpleAnnotation;

    impl fmt::Display for SimpleAnnotation {
        fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
            fmt.write_str("Type")
        }
    }

    #[test]
    fn multiple_writes() {
        let annotated = Annotated {
            content: &MultipleWrites,
            annotation: &SimpleAnnotation,
        };

        let output = format!("{annotated}");
        let expected = "┌─ foobarhello\n│  world\n└→ Type";

        assert_eq!(output, expected);
    }

    #[test]
    fn proper_newline_handling() {
        let annotated = Annotated {
            content: &"first line\nsecond line",
            annotation: &"annotation",
        };

        let output = format!("{annotated}");
        let expected = "┌─ first line\n│  second line\n└→ annotation";

        assert_eq!(output, expected);
    }

    #[test]
    fn single_line_content() {
        let annotated = Annotated {
            content: &"single line",
            annotation: &"type info",
        };

        let output = format!("{annotated}");
        let expected = "┌─ single line\n└→ type info";

        assert_eq!(output, expected);
    }
}
