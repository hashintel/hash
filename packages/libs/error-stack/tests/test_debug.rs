#![cfg_attr(nightly, feature(provide_any))]
// can be considered safe, because we only check the output, which in itself does not use **any**
// unsafe code.
#![cfg(not(miri))]
#![cfg_attr(all(nightly, feature = "std"), feature(error_generic_member_access))]

mod common;
use common::*;
#[allow(unused_imports)]
use error_stack::Report;
use insta::assert_snapshot;
#[cfg(feature = "spantrace")]
use tracing_error::ErrorLayer;
#[cfg(feature = "spantrace")]
use tracing_subscriber::layer::SubscriberExt;

#[cfg(feature = "spantrace")]
fn setup_tracing() {
    static ONCE: std::sync::Once = std::sync::Once::new();
    ONCE.call_once(|| {
        tracing::subscriber::set_global_default(
            tracing_subscriber::Registry::default().with(ErrorLayer::default()),
        )
        .expect("Could not set tracing subscriber");
    })
}

#[cfg(not(feature = "spantrace"))]
fn setup_tracing() {}

#[cfg(not(all(rust_1_65, feature = "std")))]
fn setup_backtrace() {
    std::env::set_var("RUST_LIB_BACKTRACE", "0");
}

#[cfg(all(rust_1_65, feature = "std"))]
fn setup_backtrace() {
    std::env::set_var("RUST_LIB_BACKTRACE", "1");
}

#[cfg(feature = "pretty-print")]
fn setup_color() {
    owo_colors::set_override(false);
}

#[cfg(not(feature = "pretty-print"))]
fn setup_color() {}

fn setup() {
    setup_tracing();
    setup_backtrace();
    setup_color();
}

fn snap_suffix() -> String {
    #[allow(unused_mut)]
    let mut suffix: Vec<&'static str> = vec![];

    #[cfg(feature = "spantrace")]
    {
        suffix.push("spantrace");
    }

    #[cfg(all(rust_1_65, feature = "std"))]
    if supports_backtrace() {
        suffix.push("backtrace");
    }

    #[cfg(feature = "pretty-print")]
    {
        suffix.push("pretty-print");
    }

    suffix.join("-")
}

/// Overwrite to create spantraces when necessary.
#[cfg(feature = "spantrace")]
pub fn create_report() -> Report<RootError> {
    #[tracing::instrument]
    fn func_b() -> error_stack::Result<(), RootError> {
        create_error()
    }

    #[tracing::instrument]
    fn func_a() -> error_stack::Result<(), RootError> {
        func_b()
    }

    capture_error(func_a)
}

fn prepare(suffix: bool) -> impl Drop {
    setup();

    let mut settings = insta::Settings::clone_current();
    if suffix {
        settings.set_snapshot_suffix(snap_suffix());
    }

    settings.add_filter(
        r"backtrace no\. (\d+)\n(?:  .*\n)*  .*",
        "backtrace no. $1\n  [redacted]",
    );
    settings.add_filter(
        r"span trace No\. (\d+)\n(?:  .*\n)*  .*",
        "span trace No. $1\n  [redacted]",
    );
    settings.add_filter(
        r"backtrace with( (\d+) frames)? \((\d+)\)",
        "backtrace ($3)",
    );

    settings.bind_to_scope()
}

/// Generate the `Report` for:
///
/// ```text
///                    P 1
///                     │
///                     ▼
///                    P 2
///                     │
///                     ▼
///                    C 1
///                     │
///                     ▼
///                ┌───P 3───┐
///                │         │
///                ▼         ▼
///               P 4       P 5
///                │         │
///                ▼         ▼
///               C 2      C 10
///                │         │
///                ▼         ▼
///               P 6       P 7
///                │         │
///                ▼         ▼
///          root error 1   C 3
///                          │
///                          ▼
///                         P 8
///                          │
///                          ▼
///       ┌─────────────────P 9─────────────────┬─────────────┐
///       │                  │                  │             │
///       │                  │                  │             │
///       ▼                  ▼                  ▼             ▼
///      C 4               P 16               P 11          P 12
///       │                  │                  │             │
///       ▼                  ▼                  ▼             ▼
/// root error 2       ┌───P 10───┐            C 9           C 6
///                    │          │             │             │
///                    ▼          ▼             ▼             ▼
///                  P 13       P 14      root error 5       C 7
///                    │          │                           │
///                    ▼          ▼                           ▼
///                   C 8       P 15                    root error 6
///                    │          │
///                    ▼          ▼
///              root error 3    C 5
///                               │
///                               ▼
///                         root error 4
/// ```
///
/// `P = Printable`, `C = Context`
fn create_sources_nested() -> Report<ContextA> {
    let r4 = create_report()
        .change_context(ContextA(5))
        .attach_printable("15")
        .attach_printable("14");

    let r6 = create_report()
        .change_context(ContextA(7))
        .change_context(ContextA(6))
        .attach_printable("12");

    let r5 = create_report()
        .change_context(ContextA(9))
        .attach_printable("11");

    let mut r3 = create_report()
        .change_context(ContextA(8))
        .attach_printable("13");

    r3.extend_one(r4);
    let r3 = r3.attach_printable("10").attach_printable("16");

    let mut r2 = create_report().change_context(ContextA(4));

    r2.extend_one(r3);
    r2.extend_one(r5);
    r2.extend_one(r6);

    let r2 = r2
        .attach_printable("9")
        .attach_printable("8")
        .change_context(ContextA(3))
        .attach_printable("7")
        .change_context(ContextA(10))
        .attach_printable("5");

    let mut r1 = create_report()
        .attach_printable("6")
        .change_context(ContextA(2))
        .attach_printable("4");

    r1.extend_one(r2);

    r1.attach_printable("3")
        .change_context(ContextA(1))
        .attach_printable("2")
        .attach_printable("1")
}

/// This is the main test, to test all different parts at once,
/// and demonstrates that the rendering algorithm works at arbitrary depth.
#[test]
fn sources_nested() {
    let _guard = prepare(true);

    let report = create_sources_nested();

    assert_snapshot!(format!("{report:?}"));
}

#[test]
fn sources_nested_alternate() {
    let _guard = prepare(true);

    let report = create_sources_nested();

    assert_snapshot!(format!("{report:#?}"));
}

#[cfg(all(
    rust_1_65,
    feature = "std",
    feature = "spantrace",
    feature = "pretty-print"
))]
mod full {
    //! Why so many cfg guards?
    //! What was found during initial development of the feature was,
    //! that a complete test of all tests with snapshots on every possible feature combination
    //! was infeasible, as this would lead to *a lot* of different snapshots.
    //!
    //! Changes in snapshots (this includes adding or removing lines in the test code) results in 9
    //! different snapshots, which *all* basically test the same permutation:
    //! * Does glyph/non-glyph output look nice?
    //! * Does rendering work?
    //! * Do the builtin hooks (`Backtrace` and `SpanTrace`) work?
    //!
    //! Does any combination of those work together?
    //! Therefore most of them are redundant, this means that we can cut down on the amount of
    //! snapshots that are generated.
    //! This does *not* impact speed, but makes it easier to look through all snapshots, which means
    //! that instead of 118 new snapshots once a code line changes, one just needs to look over
    //! < 30, which is a lot more manageable.
    //!
    //! There are still some big snapshot tests, which are used evaluate all of the above.

    #[cfg(nightly)]
    use std::any::Demand;
    use std::{
        error::Error,
        fmt::{Display, Formatter},
    };

    use super::*;

    #[test]
    fn linear() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable(PrintableA(0))
            .attach(AttachmentA)
            .attach(AttachmentB)
            .change_context(ContextA(0))
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn linear_ext() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable(PrintableA(0))
            .attach(AttachmentA)
            .attach(AttachmentB)
            .change_context(ContextA(0))
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:#?}"));
    }

    #[derive(Debug)]
    struct ContextC;

    impl Display for ContextC {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            f.write_str("A multiline\ncontext that might have\na bit more info")
        }
    }

    impl Error for ContextC {}

    #[test]
    fn multiline_context() {
        let _guard = prepare(false);

        let report = Report::new(ContextC)
            .change_context(ContextC)
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:#?}"));
    }

    #[test]
    fn multiline() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable("A multiline\nattachment\nthat might have some\nadditional info")
            .attach_printable("A multiline\nattachment\nthat might have some\nadditional info");

        assert_snapshot!(format!("{report:#?}"));
    }

    /// Generate the `Debug` for
    ///
    /// ```text
    ///         [A] B
    ///            |
    ///         [C] A
    ///            |
    ///         [A] A
    ///      /     |    \
    ///  [P] A  [P] B  [P] B
    ///    |       |      |
    ///   Root   Root    Root
    /// ```
    ///
    /// This should demonstrate that we're able to generate with multiple groups at the same time.
    #[test]
    fn sources() {
        let _guard = prepare(false);

        let mut root1 = create_report().attach_printable(PrintableA(1));
        let root2 = create_report().attach_printable(PrintableB(2));
        let root3 = create_report().attach_printable(PrintableB(3));

        root1.extend_one(root2);
        root1.extend_one(root3);

        let report = root1
            .attach(AttachmentA(1))
            .change_context(ContextA(2))
            .attach(AttachmentB(2));

        assert_snapshot!(format!("{report:?}"));
    }

    /// Generate the `Debug` for:
    ///
    /// ```text
    ///         [A] B
    ///            |
    ///         [C] A
    ///            |
    ///         [A] A
    ///           /  \
    ///       [P] A  [A] A
    ///         |       |   \
    ///        Root  [P] B  [P] B
    ///                 |      |
    ///               Root    Root
    /// ```
    ///
    /// and should demonstrate that "transparent" groups, groups which do not have a change in
    /// context, are still handled gracefully.
    #[test]
    fn sources_transparent() {
        let _guard = prepare(false);

        let report = {
            let mut report = create_report().attach_printable(PrintableA(1));

            report.extend_one({
                let mut report = create_report().attach_printable(PrintableB(2));

                report.extend_one(create_report().attach_printable(PrintableB(3)));

                report.attach_printable(PrintableA(4))
            });

            report
                .attach(AttachmentA(1))
                .change_context(ContextA(2))
                .attach(AttachmentB(2))
        };

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn complex() {
        let _guard = prepare(false);

        let mut report = create_report().attach_printable(PrintableA(0));
        report.extend_one({
            let mut report = create_report().attach_printable(PrintableB(1));

            report.extend_one(
                create_report()
                    .attach(AttachmentB(0))
                    .attach(AttachmentA(1))
                    .attach_printable(PrintableB(1)),
            );

            report.attach(AttachmentA(2)).attach_printable("Test")
        });

        // force the generation of a tree node
        let report = report
            .change_context(ContextA(2))
            .attach_printable(PrintableA(2));

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<u32>(|_, context| {
            context.push_body("unsigned 32bit integer");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_context() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.increment_counter();
            context.push_body(format!("unsigned 32bit integer (No. {idx})"));
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_for_context() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<RootError>(|_, _| {
            // This should not be displayed as `RootError` is only used as `Context`, never as
            // attachment.
            unreachable!("A context should never be used as hook");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_multiple() {
        let _guard = prepare(false);

        let report = create_report().attach(1u32).attach(2u64);

        Report::install_debug_hook::<u32>(|_, context| {
            context.push_body("unsigned 32bit integer");
        });
        Report::install_debug_hook::<u64>(|_, context| {
            context.push_body("unsigned 64bit integer");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_decr() {
        let _guard = prepare(false);

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.decrement_counter();
            context.push_body(idx.to_string());
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_incr() {
        let _guard = prepare(false);

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.increment_counter();
            context.push_body(idx.to_string());
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_alternate() {
        let _guard = prepare(false);

        let report = create_report().attach(2u64);

        Report::install_debug_hook::<u64>(|_, context| {
            if context.alternate() {
                context.push_appendix("Snippet");
            }

            context.push_body("Empty");
        });

        assert_snapshot!("norm", format!("{report:?}"));

        assert_snapshot!("alt", format!("{report:#?}"));
    }

    #[cfg(nightly)]
    #[derive(Debug)]
    struct ContextD {
        code: usize,
        reason: &'static str,
    }

    #[cfg(nightly)]
    impl Display for ContextD {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            f.write_str("context D")
        }
    }

    #[cfg(nightly)]
    impl Error for ContextD {
        fn provide<'a>(&'a self, req: &mut Demand<'a>) {
            req.provide_ref(&self.code);
            req.provide_ref(&self.reason);
        }
    }

    #[test]
    #[cfg(nightly)]
    fn hook_provider() {
        let _guard = prepare(false);

        let report = create_report().change_context(ContextD {
            code: 420,
            reason: "Invalid User Input",
        });

        Report::install_debug_hook::<usize>(|value, context| {
            context.push_body(format!("usize: {value}"));
        });
        Report::install_debug_hook::<&'static str>(|value, context| {
            context.push_body(format!("&'static str: {value}"));
        });

        assert_snapshot!(format!("{report:?}"));
    }
}
