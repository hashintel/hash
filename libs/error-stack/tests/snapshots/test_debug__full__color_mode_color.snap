---
source: tests/test_debug.rs
expression: "format!(\"{report:?}\")"
---
[1mroot error[22m
[31m├╴[39mat [90mtests/common.rs:9:5[39m
[31m├╴[39mbacktrace (1)
[31m╰╴[39mspan trace with 2 frames (1)

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

backtrace no. 1
  [redacted]

span trace No. 1
  [redacted]
